enhancements/insights/conditional-data-gathering.md

```diff
@@ -0,0 +1,237 @@
+- There can be up to 8 gathering functions in each rule
+- Namespace from `gather_logs_of_namespace` function should be a string of length between 1 and 128,
+match the next regular expression `^[a-zA-Z]([a-zA-Z0-9\-]+[\.]?)*[a-zA-Z0-9]$` and start with `openshift-`
+- Tail lines from `gather_logs_of_namespace` function should be an integer between 1 and 4096
+- Namespace from `gather_image_streams_of_namespace` function should be a string of length between 1 and 128,
+match the next regular expression `^[a-zA-Z]([a-zA-Z0-9\-]+[\.]?)*[a-zA-Z0-9]$` and start with `openshift-`
+
+If anything fails to be validated, the config is rejected and an error is written to the logs.
+The PR with validation on operator side - https://github.com/openshift/insights-operator/pull/483
+
+5. Insights Operator goes rule by rule and collects the requested data if corresponding conditions are met
+
+The new gathering functions and conditions are added the same way as any other code changes to insights-operator.
+The JSON schema also lives in the operator repo and is changed through the regular process.
+
+### User Stories
+
+For OpenShift engineers it will be easier to get necessary information when some problem occurs (like alert is firing).
+Getting such information now means changing operator's code which is especially painful with older versions of clusters.
+When the data is not needed anymore, the rule can be simply removed from the repo.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Empty
+
+### Risks and Mitigations
+
+The potential risks could come from an attacker spoofing the config somehow (if they got access to the repo),
+all they could do is let the insights operator collect more data and send it to the c.r.c, but because of validation
+on the operator side, the potential of collecting data is limited and there still would be
+the same anonymization of potentially sensitive data as before.
+For example, we check that namespaces start with `openshift-`, we're also limiting
+the amount of potentially collected data by, for example, introducing the limit for amount of collected logs
+(per container and the number of containers) and, in the worst case, if the conditional gathering takes too much time
+it would be stopped by the timeout and we would just get less data.
+
+Also in the regular workflow, changing the config involves going through the repo's CI (JSON schema validator)
+and probably a simple review process.
+
+Not really a risk, but in case the service is not available or provides invalid config for the operator,
+we would just have less data in the archives (everything except conditional gatherer's data),
+an error would be written to the logs and to the metadata which would allow us to know
+that something is broken and we need to bring the service back.
+
+In case GitHub is not available, we won't be able to add new rules, but the old ones would still be returned by
+the service.
+
+## Design Details
+
+Empty
```

Comment:
The total cluster impact (cpu consumed by operator, api, nodes) as a result of this gather should be approximately constant as a function of the number and type of input rules, to prevent scenarios where a mis configuration creates an expensive gather set.  I recommend rate limiting the gathering process for this component so that no more than X impact can be done in Y time regardless of the number or complexity of rules.  I would also like to see design points on the estimated cost of the gather per rule / rule type relative to other known costs within the operator.

Reply:
Besides the limit for the number of rules we also have timeout for each gatherer, so if it takes more than `GatheringInterval / 2`, the gatherer stops collecting the data. For the most of clusters which have gathering interval of 2 hours the timeout is gonna be 1 hour.